res/controllers/Denon-MC7000-scripts.js

```diff
@@ -0,0 +1,839 @@
+/**
+ * Denon DJ MC7000 DJ controller script for Mixxx 2.2.3
+ * 
+ * Started in Dec. 2019 by OsZ
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU General Public License
+ * as published by the Free Software Foundation; either version 2
+ * of the License, or (at your option) any later version.
+ * 
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ * 
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ * 
+ * Before using the mapping please make sure your MC7000 controller works for
+ * your operating system. For Windows you need driver software by Denon, Mac users 
+ * should be lucky as it shall work out-of-the-box. Linux users need to compile 
+ * their own Kernel with modified /sound/usb/clock.c see the "Denon MC7000 Mapping"
+ * thread at https://www.mixxx.org/forums/viewtopic.php?f=7&t=13126
+**/
+
+var MC7000 = {};
+
+/*///////////////////////////////////
+//      USER VARIABLES BEGIN       //
+///////////////////////////////////*/
+
+// Wanna have Needle Search active while playing a track ?
+// In any case Needle Search is available holding "SHIFT" down.
+// can be true or false (recommended: false)
+MC7000.needleSearchPlay = false;
+
+// Pitch Fader ranges to cycle through with the "RANGE" buttons.
+var lowest = 4   ;// lowest value in % (default: 4)
+var low = 6      ;// next value in % (default: 6)
+var middle = 10  ;// next value in % (default: 10)
+var high = 16    ;// next value in % (default: 16)
+var highest = 24 ;// highest value in % (default: 24)
+
+// Platter Ring LED mode
+// Mode 0 = Single "off" LED chase (all others "on")
+// Mode 1 = Single "on" LED chase (all others "off")
+// use "SHIFT" + "DECK #" to toggle between both modes
+MC7000.modeSingleLED = 1    ;// default: 1
+
+// Set Vinyl Mode on ("true") or off ("false") when MIXXX starts.
+// This sets the Jog Wheel touch detection / Vinyl Mode
+// and the Jog LEDs ("VINYL" on = spinny, "VINYL" off = track position).
+MC7000.VinylModeOn = true    ;// default: true
+
+// Scratch algorithm parameters
+MC7000.scratchParams = {
+    recordSpeed: 33.3   ,// default: 33.3
+    alpha: (1.0/10)     ,// default: (1.0/10)
+    beta: (1.0/10)/32    // default: (1.0/10)/32
+};
+
+// Sensitivity of the jog wheel (also depends on audio latency)
+// lower values make it less, higher value more sensible
+MC7000.jogParams = {
+    jogSensitivity: 30  ,// default: 30
+    maxJogValue: 3      ,// default: 3
```

Comment:
Whats the reason for this `maxJogValue`?

Reply:
Using the scratchingExample.js code from the WIKI was not satisfying for me. But as I am still pleased with the MC4000 way how the scratching feels like then I copied the formula from there. Honestly - without really knowing what the parameters tell me. Experimenting with the values showed me different behavior though.

Reply:
The parameters only change things when moving the wheel while having vinyl mode deactivated. So they shouldn't influence the Jog behavior in any way.

Reply:
Sorry for confusion: You are right - this is for controlling how sensitive the the platter is when beat matching - accidentally wrote scratching...

Reply:
I get the reason for having `jogSensitivity` but not really for `maxJogValue` because that just clamps the value given to mixxx which doesn't really make sense. It should make a difference for you at higher speeds. Can you set it to a really high value and then beatmatch a bit? does it get more difficult or can you do it faster (since not clamping the value should allow you to do faster adjustments)?

Reply:
This didn't let me go so I checked in more detail what the author of the MC4000 wanted to happen: The "jog" values generally can be set from -3 to 3. Inside the MC7000.wheelTurn function the "jog" parameter can be narrowed down in the way when it asks for the Math.min(MC7000.jogParams.maxJogValue, jogAbsolute). So even if the jogSensitivity is high, e.g you turn too hard - which influences the maxJogValue - the value for "jog" maybe maximum 3 (or lower if you set MC7000.jogParams.maxJogValue accordingly) if the variable "jogAbsolute" exceeding it. So it works like a limiter on the "jog" value.

Reply:
Yes. exactly. But I don't see a practical reason for implementing this.